pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/SnapshotSegmentAbortedTxnProcessorImpl.java

```diff
@@ -0,0 +1,646 @@
+                                                    transactionBufferSnapshotIndex));
+                            this.unsealedAbortedTxnIdSegment = (ArrayList<TxnIDData>) persistentSnapshotIndexes
+                                    .getSnapshot().getAborts();
+                            this.maxReadPosition = new PositionImpl(persistentSnapshotIndexes
+                                    .getSnapshot().getMaxReadPositionLedgerId(),
+                                    persistentSnapshotIndexes.getSnapshot().getMaxReadPositionEntryId());
+                            if (indexes.size() != 0) {
+                                persistentWorker.sequenceID.set(indexes.lastEntry().getValue().sequenceID + 1);
+                            }
+                        }
+                        //Read snapshot segment to recover aborts.
+                        ArrayList<CompletableFuture<Void>> completableFutures = new ArrayList<>();
+                        AtomicLong invalidIndex = new AtomicLong(0);
+                        AsyncCallbacks.OpenReadOnlyManagedLedgerCallback callback = new AsyncCallbacks
+                                .OpenReadOnlyManagedLedgerCallback() {
+                            @Override
+                            public void openReadOnlyManagedLedgerComplete(ReadOnlyManagedLedgerImpl readOnlyManagedLedger, Object ctx) {
+                                persistentSnapshotIndexes.getIndexList().forEach(index -> {
+                                    CompletableFuture<Void> completableFuture1 = new CompletableFuture<>();
+                                    completableFutures.add(completableFuture1);
+                                    readOnlyManagedLedger.asyncReadEntry(
+                                            new PositionImpl(index.getPersistentPositionLedgerID(),
+                                                    index.getPersistentPositionEntryID()),
+                                            new AsyncCallbacks.ReadEntryCallback() {
+                                                @Override
+                                                public void readEntryComplete(Entry entry, Object ctx) {
+                                                    //Remove invalid index
+                                                    if (entry == null) {
+                                                        indexes.remove(new PositionImpl(
+                                                                index.getMaxReadPositionLedgerID(),
+                                                                index.getMaxReadPositionEntryID()));
+                                                        completableFuture1.complete(null);
+                                                        invalidIndex.getAndIncrement();
+                                                        return;
+                                                    }
+                                                    handleSnapshotSegmentEntry(entry);
```

Comment:
Here all `snapshot-segment` are loaded into memory, our original design should be loaded on demand. Should we left a `TODO` for later optimization?